### PR TITLE
Add support for TLS (conformance test passes)

### LINKED ISSUE #199: Add TestIngressTLS test
- Goal: Pass the test which is same with `TestIngressTLS` in networking repo by conformance test and controller e2e test.
- Non-goal: Support of net-certmanager, net-http01 functionality.

HTTPRoute does not support `tls` setting on v1alpha2 (v0.4.0) so we need to use `tls.certificateRefs` in Gateway like [this example](https://github.com/kubernetes-sigs/gateway-api/blob/25607266669cc3616ed90404db4ff6e6d854a3b7/examples/v1alpha2/tls-cert-cross-namespace.yaml#L13-L18).

#### Current blocker:
- [x] https://github.com/istio/istio/issues/36075
- [ ] https://github.com/knative/pkg/pull/2327

### LINKED ISSUE #319: Clean up Gateway Spec.Listeners
There are two cases where we might remove `Spec.Listeners` entries from a `Gateway` resource:

1. The KIngress associated with the Listener has been deleted.
2. The KIngress associated with the Listener has been changed to have (for example) a different hostname.

Right now, we don't handle either of these scenarios.

Scenario 2 may be somewhat harder; for case 1 we could use a finalizer to make sure that the Gateway entries are removed before the KIngress resource is deleted.

Another option is to turn the reconciliation around and reconcile the `Gateway` resource(s) across all the KIngresses. I'm not doing that right now because that's not the way the current code is structured, but given #318, that may be a good option.

### PR DESCRIPTION
**Pull request by evankanderson**
# Changes

- :bug: Add support for TLS by filling out Gateway's `Spec.Listener`
- :broom: Fix a few typo'ed import aliases

Still TODO:
- Clean up `Spec.Listener` on delete (separate PR)
- More testing in `ingress_test.go` (I got one done, but I want to test the "listener already exists" scenario, at least)

/kind enhancement

Fixes #199
Fixes #319

**Release Note**

```release-note
Adds support for Knative-configured TLS ("auto-TLS")
```

**Docs**

```docs
This increases parity with other implementations, but should not require its own
documentation, as no configuration options were added.
```

### CODE REVIEW

File: pkg/reconciler/ingress/reconcile_resources.go [hunk 1]
```diff
@@ -82,3 +85,156 @@ func (c *Reconciler) reconcileHTTPRoute(
...
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      externalGw.Gateway.Name,
+			Namespace: externalGw.Gateway.Namespace,
+		},
+	}
+	secret := metav1.PartialObjectMetadata{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1", //metav1.GroupVersion.String(),
```
Comment (dprotaso) [inline pkg/reconciler/ingress/reconcile_resources.go:110]: nit: corev1.Secret not metav1

File: pkg/reconciler/ingress/reconcile_resources.go [hunk 2]
```diff
@@ -82,3 +85,156 @@ func (c *Reconciler) reconcileHTTPRoute(
...
+		return err
+	}
+
+	update := gw.DeepCopy()
+
+	lmap := map[string]*gatewayv1alpha2.Listener{}
+	for _, l := range listeners {
+		lmap[string(l.Name)] = l
+	}
+	// TODO: how do we track and remove listeners if they are removed from the KIngress spec?
```
Comment (dprotaso) [inline pkg/reconciler/ingress/reconcile_resources.go:206]: issue?

File: pkg/reconciler/ingress/reconcile_resources.go [hunk 3]
```diff
@@ -82,3 +85,156 @@ func (c *Reconciler) reconcileHTTPRoute(
...
+		// Add all remaining listeners
+		update.Spec.Listeners = append(update.Spec.Listeners, *l)
+		updated = true
+	}
+
+	if updated {
+		_, err := c.gwapiclient.GatewayV1alpha2().Gateways(update.Namespace).Update(
+			ctx, update, metav1.UpdateOptions{})
+		if err != nil {
+			recorder.Eventf(ing, corev1.EventTypeWarning, "GatewayUpdateFailed", "Failed to update Gateway %s: %v", gwName, err)
```
[the remark this answers is not in the document]
Comment (dprotaso) [reply pkg/reconciler/ingress/reconcile_resources.go:234]: Also, we shouldn't run this controller in HA with the number of buckets being greater than 1.

**Review by dprotaso — COMMENTED** [review]
whoops accidentally hit the wrong button earlier

change looks good 

tl;dr
- I think one of the lint warnings is a legit bug
- We should have a follow up where we create multiple gateways vs updating a single one

**Comment by evankanderson** [comment]
Okay, I've added a couple more test cases, going over Dave's comments now.

I think my next steps are going to be to figure out what the Contour maintainers expect from `AppPrortocol` to enable Webhook and TLS, and see whether or not I can reasonably hack that into Contour to get the `websocket` and `grpc` tests passing.

File: pkg/reconciler/ingress/reconcile_resources.go [hunk 4]
```diff
@@ -82,3 +85,156 @@ func (c *Reconciler) reconcileHTTPRoute(
...
+
+	for _, l := range lmap {
+		// Add all remaining listeners
+		update.Spec.Listeners = append(update.Spec.Listeners, *l)
+		updated = true
+	}
+
+	if updated {
+		_, err := c.gwapiclient.GatewayV1alpha2().Gateways(update.Namespace).Update(
+			ctx, update, metav1.UpdateOptions{})
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply pkg/reconciler/ingress/reconcile_resources.go:232]: https://github.com/knative-sandbox/net-gateway-api/issues/317

File: pkg/reconciler/ingress/reconcile_resources.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (evankanderson) [reply pkg/reconciler/ingress/reconcile_resources.go:234]: Ah, I see what you're suggesting. I'll add o TODO and a comment here.

I'll also bring this up with the Gateway API maintainers -- it feels like generating hundreds of Gateways that are supposed to collapse onto the same proxy is probably not right (and doesn't seem to work for at least Contour today), but the one-Gateway model doesn't scale, either.

Filed https://github.com/knative-sandbox/net-gateway-api/issues/318

File: pkg/reconciler/ingress/ingress.go [hunk 5]
```diff
@@ -100,14 +105,33 @@ func (c *Reconciler) reconcileIngress(ctx context.Context, ing *v1alpha1.Ingress
...
+		if err != nil {
+			return err
+		}
+		listeners = append(listeners, l...)
+	}
+
+	if len(listeners) > 0 {
+		// For now, we only reconcile the external visibility, because there's
+		// no way to provide TLS for internal listeners.
+		err := c.reconcileGatewayListeners(
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply pkg/reconciler/ingress/ingress.go:122]: I'm not sure -- how would you decide which to do?

File: pkg/reconciler/ingress/reconcile_resources.go (hunk 2 above)
Comment (evankanderson) [reply pkg/reconciler/ingress/reconcile_resources.go:206]: https://github.com/knative-sandbox/net-gateway-api/issues/319

File: pkg/reconciler/ingress/ingress_test.go [hunk 6]
```diff
@@ -28,6 +28,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	clientgotesting "k8s.io/client-go/testing"
+	"k8s.io/utils/pointer"
```
Comment (dprotaso) [inline pkg/reconciler/ingress/ingress_test.go:31]: nit:` knative/pkg/ptr`

**Comment by evankanderson** [comment]
/hold cancel

Gateway merging doesn't actually work for at least Contour, and if you have two Gateways, there's no clear way to determine which Gateway the ones without addresses merge to.

**Comment by evankanderson** [comment]
I added some more code, we now have distinctly-named listeners of the form `kni-$INGRESS_UID`, and a finalizer to clean them up.

/assign @carlisia 

@dprotaso @nak3 not sure if you want to take (another?) look.

**Comment by nak3** [comment]
/test integration-tests_net-gateway-api_main

LGTM. Thank you!

File: pkg/reconciler/ingress/ingress_test.go (hunk 6 above)
Comment (evankanderson) [reply pkg/reconciler/ingress/ingress_test.go:31]: Why prefer `knative.dev` packages to the upstream k8s ones?

**Review by evankanderson — COMMENTED** [review]
Missed one comment earlier.

Also, this is now merged with #315, and unit tests should be passing. I'm checking e2e now, but this is ready for review.

@carlisia @dprotaso @nak3 

I wouldn't mind this getting in before nightly testing / as a candidate for 1.6...

**Comment by evankanderson** [comment]
Ugh, I've somehow screwed up _just_ the kind e2e tests...

**Comment by evankanderson** [comment]
I missed the rename `ReferencePolicy` -> `ReferenceGrant` in the ClusteRole for the controller, so it couldn't manage the resources it thought it needed to create.

/hold cancel

**Comment by evankanderson** [comment]
Tests run locally seem to pass now... (I had a cleanup issue I was debugging when I pushed before the `/hold`, so I figured I'd get the CI tests in parallel).

**Comment by evankanderson** [comment]
Hmm, it looks like the conformance tests in the e2e tests are unhappy; I can repro locally with Istio, but not with Contour. Debugging...

**Comment by carlisia** [comment]
The PR changes lgtm. I'll standby...

File: pkg/reconciler/ingress/resources/httproute.go [hunk 7]
```diff
@@ -75,6 +75,8 @@ func makeHTTPRouteSpec(
 	namespacedNameGateway := gatewayConfig.Gateways[rule.Visibility].Gateway
 
 	gatewayRef := gatewayv1alpha2.ParentReference{
+		Group:     (*gatewayv1alpha2.Group)(&gatewayv1alpha2.GroupVersion.Group),
+		Kind:      (*gatewayv1alpha2.Kind)(pointer.String("Gateway")),
```
Comment (carlisia) [inline pkg/reconciler/ingress/resources/httproute.go:79]: Does the `Port` not need to be added as well?

**Comment by carlisia** [comment]
Whoa side note: I was wondering where and how `ISTIO_VERSION` was being used since I couldn't find it anywhere in the code base. It is used by the Istio script for its installation ([https://raw.githubusercontent.com/istio/istio/master/release/downloadIstioCtl.sh](https://raw.githubusercontent.com/istio/istio/master/release/downloadIstioCtl.sh)). 

Here are the changes to the latest version, v1.14.1: [Istio / Announcing Istio 1.14.1](https://istio.io/latest/news/releases/1.14.x/announcing-1.14.1/). I wonder if this update would help with the tls stuff.

PS: I'll document the info about the Istio version.

**Comment by carlisia** [comment]
Also: Istio has not had a release since upgrading their code to use v0.5.0-rc1 ([gateway-api: upgrade to v0.5.0-rc1 by howardjohn · Pull Request #39506 · istio/istio](https://github.com/istio/istio/pull/39506?w=1)).

**Comment by carlisia** [comment]
Trying to update Istio is failing: [Bump Istio to v1.14.1 by carlisia · Pull Request #330 · knative-sandbox/net-gateway-api](https://github.com/knative-sandbox/net-gateway-api/pull/330?w=1)

**Comment by evankanderson** [comment]
I foolishly attempted to use the new `ReferenceGrant` API immediately; I've since mended my ways and switched back to using ReferencePolicy.

Pro tip when doing a merge like this:

1. Make sure the e2e test passes **before** upgrade
2. Upgrade the code and run the e2e test _on the pre-merge controller_ with _post-merge tests_.
3. Redeploy the controller and verify the e2e tests continue to pass.

Once I did this, triage of the tests was a lot easier, because it was clear that it was something in the controller code, not the merge, that had broken the tests.

**Comment by evankanderson** [comment]
Tests, including TLS tests, now pass locally.

**Comment by carlisia** [comment]
This is great, thank you! 🎉

I will add an issue to switch to `ReferenceGrant` once the providers are updated to support it. And I'll add your tips to the dev documentation.

Thanks for creating all the issues for followup work. Wanted to surface this one comment, I think it requires an issue as well? https://github.com/knative-sandbox/net-gateway-api/pull/316?w=1#discussion_r909781272

/lgtm
/approve

🚀
